README
This GStreamer plugin takes a stereo signal and puts it into two specified
channels of a multichannel audio stream. The other channels will remain
silent.

You can use this plugin to send multiple stereo streams to a surround sound
card (when you use it in combination with the adder plugin).

Usage example:
  $ gst-launch -v -m audiotestsrc ! rearrange channels=4 pos=1 ! alsasink
    Sends a sine wave to the rear channels of a four-channel alsa stream

Properties:
- channels: Specifies the output channel count.
  This can be an integer in the range of 2..8
- pos: Specifies the output channels where the plugin should put the original
  input signal. It accepts the following values:
  0: front channels
  1: rear channels
  2: center/lfe
  3: side channels

If _channels_ < 2*(pos+1), the output signal will be silent (you can use
this to generate a multichannel silence if you need it)


This code is provided under a MIT license [1], which basically means "do
with it as you wish, but don't blame us if it doesn't work". You can use
this code for any project as you wish, under any license as you wish. We
recommend the use of the LGPL [2] license for applications and plugins,
given the minefield of patents the multimedia is nowadays. See our website
for details [3].

Build this plugin like this:

    ./autogen.sh
    make

Once the plugin is built you can either install it with 'sudo make install'
(however, this will by default go into the /usr/local prefix where it won't
be picked up by a GStreamer installed from packages, so you would need to
set the GST_PLUGIN_PATH environment variable to include or point to
/usr/local/lib/gstreamer-0.10/ for your plugin to be found by a from-package
GStreamer). Alternatively, you will find your plugin binary in
gst-plugins/src/.libs/ as libgstplugin.so or similar (the extension may vary),
so you can also set the GST_PLUGIN_PATH environmen variable to the
gst-plugins/src/.libs/ directory (best to specify an absolute path though).

You can also check if it has been built correctly with:

    gst-inspect-0.10 gst-plugins/src/.libs/libgstplugin.so

This Plugin was built upon the generic GStreamer plugin template found at [4].


[1] http://www.opensource.org/licenses/mit-license.php or COPYING.MIT
[2] http://www.opensource.org/licenses/lgpl-license.php or COPYING.LIB
[3] http://gstreamer.freedesktop.org/documentation/licensing.html
[4] http://gstreamer.freedesktop.org/data/doc/gstreamer/head/pwg/html/chapter-building-boiler.html


debian/changelog
gstreamer0.10-rearrange (0.3-1) stable; urgency=low

  * Fixed some memory issues

 -- Manuel Reithuber <manuel@reithuber.net>  Tue, 24 Jul 2012 10:36:40 +0200

gstreamer0.10-rearrange (0.3) stable; urgency=low

  * Added a simple mode that just resets the channel positions
    (which is the new default mode)

 -- Manuel Reithuber <manuel@reithuber.net>  Fri, 18 May 2012 13:53:27 +0200

gstreamer0.10-rearrange (0.2-4) stable; urgency=low

  * removed dh-autoreconf from the build dependencies
    (as it's not yet present in some older distros)

 -- Manuel Reithuber <manuel@reithuber.net>  Thu, 10 May 2012 10:48:05 +0200

gstreamer0.10-rearrange (0.2-3) stable; urgency=low

  * Changed to "3.0 (native)" source format

 -- Manuel Reithuber <manuel@reithuber.net>  Thu, 10 May 2012 10:38:41 +0200

gstreamer0.10-rearrange (0.2-2) stable; urgency=low

  * debhelper downgrade (8 => 7) to support older systems

 -- Manuel Reithuber <manuel@reithuber.net>  Tue, 13 Mar 2012 13:54:20 +0100

gstreamer0.10-rearrange (0.2-1) stable; urgency=low

  * build-dependency update

 -- Manuel Reithuber <manuel@reithuber.net>  Mon, 12 Mar 2012 13:52:17 +0100

gstreamer0.10-rearrange (0.2) stable; urgency=low

  * mono support
  * proper multichannel support
  * memory leak fixes 

 -- Manuel Reithuber <manuel@reithuber.net>  Mon, 12 Mar 2012 13:47:56 +0100

gstreamer0.10-rearrange (0.1-2) stable; urgency=low

  * updated some debian stuff

 -- Manuel Reithuber <manuel@reithuber.net>  Thu, 08 Mar 2012 17:50:49 +0100

gstreamer0.10-rearrange (0.1-1) stable; urgency=low

  * Initial release

 -- Manuel Reithuber <manuel@reithuber.net>  Thu, 08 Mar 2012 17:50:22 +0100


src/gstrearrange.h
/*
 * GStreamer
 * Copyright (C) 2005 Thomas Vander Stichele <thomas@apestaart.org>
 * Copyright (C) 2005 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright (C) 2012 Manuel Reithuber <gstreamer@manuel.reithuber.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef __GST_REARRANGE_H__
#define __GST_REARRANGE_H__

#include <gst/gst.h>

G_BEGIN_DECLS

/* #defines don't like whitespacey bits */
#define GST_TYPE_REARRANGE \
  (gst_rearrange_get_type())
#define GST_REARRANGE(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_REARRANGE,GstReArrange))
#define GST_REARRANGE_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_REARRANGE,GstReArrangeClass))
#define GST_IS_REARRANGE(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_REARRANGE))
#define GST_IS_REARRANGE_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_REARRANGE))

typedef struct _GstReArrange      GstReArrange;
typedef struct _GstReArrangeClass GstReArrangeClass;

enum EReArrangeMode {
	REARRANGE_MODE_SIMPLE = 0,
	REARRANGE_MODE_SPREAD = 1
};

struct _GstReArrange {
  GstElement element;

  GstPad *sinkpad, *srcpad;

  guint outChannels;
  guint outPos;
  enum EReArrangeMode mode;
};

struct _GstReArrangeClass
{
  GstElementClass parent_class;
};

GType gst_rearrange_get_type (void);

G_END_DECLS

#endif /* __GST_REARRANGE_H__ */


src/gstrearrange.c
/*
 * GStreamer
 * Copyright (C) 2005 Thomas Vander Stichele <thomas@apestaart.org>
 * Copyright (C) 2005 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright (C) 2012 Manuel Reithuber <gstreamer@manuel.reithuber.net>>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/**
 * SECTION:element-rearrange
 *
 * Rearrange spreads two input channels to two channels of a 5.1 or 7.1 signal.
 * It therefore allows you to send more than one stereo signals to multichannel
 * sound cards.
 *
 * Keep in mind that some audio backends do up-/downmixing if the signal's
 * channel count doesn't match the channel count of the audio device
 *
 * <refsect2>
 * <title>Example launch lines</title>
 * |[
 * gst-launch -v -m audiotestsrc ! rearrange channels=4 pos=1 ! alsasink
 * ]|
 * Plays a sine wave to the rear channels of a 4-channel output stream
 *
 * |[
 * gst-launch audiotestsrc freq=1200 ! rearrange name=foo channels=4 ! adder name=mix ! alsasink  audiotestsrc freq=444 ! rearrange name=bar channels=4 pos=1 ! mix.
 * ]|
 * Plays a 1200Hz wave at the front and a 444Hz sine at the rear channels of a 4-channel output stream
 *
 * </refsect2>
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <gst/gst.h>
#include <gst/audio/multichannel.h>

#include "gstrearrange.h"

GST_DEBUG_CATEGORY_STATIC (gst_rearrange_debug);
#define GST_CAT_DEFAULT gst_rearrange_debug

/* Filter signals and args */
enum {
	/* FILL ME */
	LAST_SIGNAL
};

enum {
	PROP_0,
	PROP_POS,
	PROP_CHANNELS,
	PROP_MODE
};

#define SRCCAPS \
	"audio/x-raw-int," \
		"width = (int) 16," \
		"depth = (int) 16," \
		"endianness = (int) BYTE_ORDER," \
		"channels = (int) {2,4,6,8}," \
		"rate = (int) [1,2147483647];" \
	"audio/x-raw-float," \
		"width = (int) {32, 64}," \
		"endianness = (int) BYTE_ORDER," \
		"channels = (int) {2,4,6,8}," \
		"rate = (int) [1,2147483647]"


#define SINKCAPS \
	"audio/x-raw-int," \
		"width = (int) 16," \
		"depth = (int) 16," \
		"endianness = (int) BYTE_ORDER," \
		"channels = (int) {1,2}," \
		"rate = (int) [1, 2147483647];" \
	"audio/x-raw-float," \
		"width = (int) {32, 64}," \
		"endianness = (int) BYTE_ORDER," \
		"channels = (int) {1,2}," \
		"rate = (int) [1,2147483647]"

static GstAudioChannelPosition _gst_rearrange_positions[8] = {
	GST_AUDIO_CHANNEL_POSITION_FRONT_LEFT,
	GST_AUDIO_CHANNEL_POSITION_FRONT_RIGHT,
	GST_AUDIO_CHANNEL_POSITION_REAR_LEFT,
	GST_AUDIO_CHANNEL_POSITION_REAR_RIGHT,
	GST_AUDIO_CHANNEL_POSITION_FRONT_CENTER,
	GST_AUDIO_CHANNEL_POSITION_LFE,
	GST_AUDIO_CHANNEL_POSITION_SIDE_LEFT,
	GST_AUDIO_CHANNEL_POSITION_SIDE_RIGHT
};

/* the capabilities of the inputs and outputs.
 *
 * describe the real formats here.
 */
static GstStaticPadTemplate sink_factory = GST_STATIC_PAD_TEMPLATE ("sink",
	GST_PAD_SINK,
	GST_PAD_ALWAYS,
	GST_STATIC_CAPS(SINKCAPS)
	);

static GstStaticPadTemplate src_factory = GST_STATIC_PAD_TEMPLATE ("src",
	GST_PAD_SRC,
	GST_PAD_ALWAYS,
	GST_STATIC_CAPS(SRCCAPS)
	);

GST_BOILERPLATE (GstReArrange, gst_rearrange, GstElement, GST_TYPE_ELEMENT);

static void gst_rearrange_set_property (GObject * object, guint prop_id,
	const GValue * value, GParamSpec * pspec);
static void gst_rearrange_get_property (GObject * object, guint prop_id,
	GValue * value, GParamSpec * pspec);

static gboolean gst_rearrange_set_caps (GstPad * pad, GstCaps * caps);
static GstFlowReturn gst_rearrange_chain (GstPad * pad, GstBuffer * buf);

/* GObject vmethod implementations */

static void
gst_rearrange_base_init (gpointer gclass) {
  GstElementClass *element_class = GST_ELEMENT_CLASS (gclass);

  gst_element_class_set_details_simple(element_class,
	"ReArrange",
	"Re-arrange audio channels",
	"Move the two channels of an incoming stereo signal to two of the other channels (e.g. rear, side or center/lfe)",
	"Manuel Reithuber <gstreamer@manuel.reithuber.net>");

  gst_element_class_add_pad_template (element_class,
	  gst_static_pad_template_get (&src_factory));
  gst_element_class_add_pad_template (element_class,
	  gst_static_pad_template_get (&sink_factory));

}

/* initialize the rearrange's class */
static void
gst_rearrange_class_init (GstReArrangeClass * klass) {
	GObjectClass *gobject_class;
//	GstElementClass *gstelement_class;

	gobject_class = (GObjectClass *) klass;
//	gstelement_class = (GstElementClass *) klass;

	gobject_class->set_property = gst_rearrange_set_property;
	gobject_class->get_property = gst_rearrange_get_property;

	g_object_class_install_property (gobject_class, PROP_CHANNELS,
		g_param_spec_uint("channels", "OutputChannels", "Channel count of the output signal",
		2, 8, 8, G_PARAM_READWRITE));
	g_object_class_install_property (gobject_class, PROP_POS,
		g_param_spec_uint("pos", "SignalPos", "Position of the signal (0: front, 1: rear, 2: center/lfe, 3: side)",
		0, 3, 0, G_PARAM_READWRITE));
	g_object_class_install_property (gobject_class, PROP_MODE,
		g_param_spec_uint("mode", "Rearrange-mode",
			"Specify the working mode for the plugin\n"
			"\t0: Simple: Just annotate the output channels with the new positions (Simpler, but won't work if the sound daemon does remixing!"
			"\t1: Spread: create [channels] channels and fill the unused ones with silence\n",
			0, 1, 0, G_PARAM_READWRITE));
}

/* initialize the new element
 * instantiate pads and add them to element
 * set pad calback functions
 * initialize instance structure
 */
static void
gst_rearrange_init (GstReArrange * filter, GstReArrangeClass * gclass) {
  filter->sinkpad = gst_pad_new_from_static_template (&sink_factory, "sink");
  gst_pad_set_setcaps_function (filter->sinkpad,
								GST_DEBUG_FUNCPTR(gst_rearrange_set_caps));
  gst_pad_set_getcaps_function (filter->sinkpad,
								GST_DEBUG_FUNCPTR(gst_pad_proxy_getcaps));
  gst_pad_set_chain_function (filter->sinkpad,
							  GST_DEBUG_FUNCPTR(gst_rearrange_chain));

  filter->srcpad = gst_pad_new_from_static_template (&src_factory, "src");
  gst_pad_set_getcaps_function (filter->srcpad,
								GST_DEBUG_FUNCPTR(gst_pad_proxy_getcaps));

  gst_element_add_pad (GST_ELEMENT (filter), filter->sinkpad);
  gst_element_add_pad (GST_ELEMENT (filter), filter->srcpad);

  filter->outChannels = 8;
  filter->outPos = 0;
  filter->mode = REARRANGE_MODE_SIMPLE;
}

static void
gst_rearrange_set_property (GObject * object, guint prop_id,
	const GValue * value, GParamSpec * pspec)
{
	GstReArrange *filter = GST_REARRANGE (object);

	switch (prop_id) {
	case PROP_CHANNELS:
		filter->outChannels = g_value_get_uint(value);
		break;
	case PROP_POS:
		filter->outPos = g_value_get_uint(value);
		break;
	case PROP_MODE:
		filter->mode = g_value_get_uint(value);
		break;
	default:
		G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
		break;
  }
}

static void
gst_rearrange_get_property (GObject * object, guint prop_id,
	GValue * value, GParamSpec * pspec)
{
	GstReArrange *filter = GST_REARRANGE (object);

	switch (prop_id) {
	case PROP_CHANNELS:
		g_print("chanType: %s\n", g_type_name(g_value_array_get_type()));
		g_value_set_uint(value, filter->outChannels);
		break;
	case PROP_POS:
		g_print("posType: %s\n", g_type_name(g_value_array_get_type()));
		g_value_set_uint(value, filter->outPos);
		break;
	case PROP_MODE:
		g_print("modeType: %s\n", g_type_name(g_value_array_get_type()));
		g_value_set_uint(value, filter->mode);
		break;
	default:
		G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
		break;
	}
}

/* GstElement vmethod implementations */

/* this function handles the link with other elements */
static gboolean
gst_rearrange_set_caps (GstPad * pad, GstCaps * caps)
{
  GstReArrange *filter;
  GstPad *otherpad;

  filter = GST_REARRANGE (gst_pad_get_parent (pad));
  otherpad = (pad == filter->srcpad) ? filter->sinkpad : filter->srcpad;
  gst_object_unref (filter);

  return gst_pad_set_caps (otherpad, caps);
}

/**
 * Copies the in buffer's caps and changes the channel parameter according
 * to the "channels" property
 * \param sinkCaps Input capabilities (those of the input buffer)
 * \param channels Output channels
 * \note this function is called by gst_rearrange_chain()
 * \warning Don't forget to call gst_caps_unref()
 */
GstCaps* gst_rearrange_set_buffer_caps (GstCaps *sinkCaps, int channels) {
	GstCaps *rc = 0;

	GstStructure *struc = gst_structure_copy(gst_caps_get_structure(sinkCaps, 0));

	gst_structure_set (struc, "channels", G_TYPE_INT, channels, NULL);

	gst_audio_set_channel_positions(struc, _gst_rearrange_positions);

	rc = gst_caps_new_empty();
	gst_caps_append_structure(rc, struc);

	//gst_structure_free(struc); Not necessary because gst_caps_append_structure makes rc the new owner
	return rc;
}

gint gst_rearrange_get_caps_int(GstCaps *sinkCaps, const char *field) {
	gint rc = 0;
	GstStructure *struc = gst_caps_get_structure(sinkCaps, 0);
	if (!gst_structure_get_int(struc, field, &rc)) {
		gchar *capString = gst_caps_to_string(sinkCaps);
		g_print("Problem getting the cap '%s' (caps: '%s')\n", field, capString);
		g_free(capString);
	}

	return rc;
}

// specific chain functions for the different working modes
inline static GstFlowReturn
gst_rearrange_chain_simple (GstReArrange *filter, GstBuffer * buf);
inline static GstFlowReturn
gst_rearrange_chain_spread (GstReArrange *filter, GstPad * pad, GstBuffer * buf);


/* chain function
 * this function does the actual processing
 */
static GstFlowReturn
gst_rearrange_chain (GstPad * pad, GstBuffer * buf) {
	GstReArrange *filter = GST_REARRANGE (GST_OBJECT_PARENT (pad));
	GstFlowReturn rc = GST_FLOW_WRONG_STATE;

	switch (filter->mode) {
	case REARRANGE_MODE_SIMPLE:
		rc = gst_rearrange_chain_simple(filter, buf);
		break;
	case REARRANGE_MODE_SPREAD:
		rc = gst_rearrange_chain_spread(filter, pad, buf);
		break;
	}
	return rc;
}

inline static GstFlowReturn
gst_rearrange_chain_simple (GstReArrange *filter, GstBuffer * buf) {
	// set channel positions
	GstCaps *tgtCaps = 0;
	GstCaps *srcCaps = gst_buffer_get_caps(buf);
	GstStructure *struc = gst_structure_copy(gst_caps_get_structure(srcCaps, 0));
	gst_caps_unref(srcCaps);

	gst_audio_set_channel_positions(struc, _gst_rearrange_positions + 2*filter->outPos);

	tgtCaps = gst_caps_new_empty();
	gst_caps_append_structure(tgtCaps, struc);

	gst_buffer_set_caps(buf, tgtCaps);
	gst_caps_unref(tgtCaps);

	//gst_structure_free(struc); Not necessary because gst_caps_append_structure makes rc the new owner

	return gst_pad_push(filter->srcpad, buf);
}

inline static GstFlowReturn
gst_rearrange_chain_spread (GstReArrange *filter, GstPad * pad, GstBuffer * buf) {
	// get input buffer info
	GstCaps *srcCaps = gst_buffer_get_caps(buf);
	int width = gst_rearrange_get_caps_int(srcCaps, "width")/8;
	int inChannels = gst_rearrange_get_caps_int(srcCaps, "channels");

	GstBuffer *tgtBuf = gst_buffer_new_and_alloc(GST_BUFFER_SIZE(buf)*(filter->outChannels/inChannels));

	// target caps
	GstCaps *tgtCaps = gst_rearrange_set_buffer_caps(srcCaps, filter->outChannels);
	gst_buffer_set_caps(tgtBuf, tgtCaps);
	gst_caps_unref(srcCaps);
	gst_caps_unref(tgtCaps);

	gint8 *pSrc = buf->data, *pTgt = tgtBuf->data;

	int curByte = 0, fromByte = 2*filter->outPos*width;
	int toByte = fromByte + (width*2); // two channels

	gint8 *pTmp = 0;
	while (pTgt < tgtBuf->data + tgtBuf->size) {
		if (curByte >= fromByte && curByte < toByte) {
			if (inChannels == 1) {
				if (curByte == fromByte) pTmp = pSrc;
				if (curByte == fromByte+width) pSrc = pTmp;
			}
			*pTgt++ = *pSrc++;
		}
		else *pTgt++ = 0;
		curByte = (curByte+1) % (filter->outChannels*width);
	}

	gst_buffer_unref(buf);
	return gst_pad_push(filter->srcpad, tgtBuf);
}


/* entry point to initialize the plug-in
 * initialize the plug-in itself
 * register the element factories and other features
 */
static gboolean
rearrange_init (GstPlugin * rearrange)
{
  /* debug category for fltering log messages
   *
   * exchange the string 'Template rearrange' with your description
   */
  GST_DEBUG_CATEGORY_INIT (gst_rearrange_debug, "rearrange",
	  0, "Template rearrange");

  return gst_element_register (rearrange, "rearrange", GST_RANK_NONE,
	  GST_TYPE_REARRANGE);
}

/* PACKAGE: this is usually set by autotools depending on some _INIT macro
 * in configure.ac and then written into and defined in config.h, but we can
 * just set it ourselves here in case someone doesn't use autotools to
 * compile this code. GST_PLUGIN_DEFINE needs PACKAGE to be defined.
 */
#ifndef PACKAGE
#define PACKAGE "rearrange"
#endif

/* gstreamer looks for this structure to register rearranges
 *
 * exchange the string 'Template rearrange' with your rearrange description
 */
GST_PLUGIN_DEFINE (
	GST_VERSION_MAJOR,
	GST_VERSION_MINOR,
	"rearrange",
	"Channel rearrangement plugin",
	rearrange_init,
	VERSION,
	"LGPL",
	"GStreamer",
	"http://gstreamer.net/"
)
